### PR TITLE
Feature: Dynamic resource notices

### PR DESCRIPTION
**Pull request by B-T-D**
### Summary of Changes

- Adds support for dynamic, programmatic notices on tables and other resources
- Organizations using Amundsen can configure a custom notice client to generate notices for tables based on arbitrary business logic. For example, at notices about delays in updates to a table's data. 

### Tests

Added unit tests for the notice endpoint.

### Documentation

Docstrings on public classes.

### CheckList
Make sure you have checked **all** steps below to ensure a timely review.
- [x] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
    - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [x] PR includes a summary of changes.
- [x] PR adds unit tests, updates existing unit tests, __OR__ documents why no test additions or modifications are needed.
- [x] In case of new functionality, my PR adds documentation that describes how to use it.
    - All the public functions and the classes in the PR contain docstrings that explain what it does

### CODE REVIEW

File: frontend/tests/unit/api/notice/test_v0.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        pass
+
+
+dummy_notice_client_class = 'tests.unit.api.notice.test_v0.DummyNoticeClient'
+
+
+class NoticeTest(unittest.TestCase):
+
+    def setUp(self) -> None:
+        local_app.config['NOTICE_CLIENT_ENABLED'] = True
```
Comment (B-T-D) [inline frontend/tests/unit/api/notice/test_v0.py:26]: get rid, don't have that in config anymore

File: frontend/tests/unit/api/notice/test_v0.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        """
+        # Reset side effects of other tests to ensure results are same regardless of execution order.
+        v0.NOTICE_CLIENT_INSTANCE = None
+
+        local_app.config['NOTICE_CLIENT'] = None
+
+        with local_app.test_client() as test:
+            response = test.get('/api/notices/v0/table?key=some_table_key')
+            self.assertEqual(response.status_code, HTTPStatus.NOT_IMPLEMENTED)
+            self.assertEqual(response.json, {'placeholder': 'this feature is not implemented'})
```
Comment (B-T-D) [inline frontend/tests/unit/api/notice/test_v0.py:41]: make this a constant so the copy can be consistent (casing, punctuation) across multiple endpoints using this boilerplate. E.g. quality also uses it.

Comment (B-T-D) [inline frontend/tests/unit/api/notice/test_v0.py:41]: Standard boilerplate key is 'msg' not 'placeholder'. See announcements and quality

Comment (B-T-D) [reply frontend/tests/unit/api/notice/test_v0.py:41]: Also can write something more verbose and helpful, announcements does "a client for retrieving announcements must be configured:

File: frontend/amundsen_application/static/js/components/Alert/Alert.tsx
Comment (kristenarmes) [inline frontend/amundsen_application/static/js/components/Alert/Alert.tsx:1]: is this file meant to replace this? https://github.com/amundsen-io/amundsen/blob/main/frontend/amundsen_application/static/js/components/Alert/index.tsx

I remember reviewing a few small changes recently that seem to have been made in the index.tsx file but not here. not sure if this one just hasn't been updated, and if so should the index file be deleted as part of this PR?

File: frontend/amundsen_application/api/notice/v0.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+
+@notices_blueprint.route('/table', methods=['GET'])
+def get_table_notices_summary() -> Response:
+    global NOTICE_CLIENT_INSTANCE
+    try:
+        client = get_notice_client()
+        if client is not None:
+            return _get_table_notices_summary_client()
+        payload = jsonify({'placeholder': 'this feature is not implemented'})
```
Comment (kristenarmes) [inline frontend/amundsen_application/api/notice/v0.py:40]: for consistency, to have the same payload format as the other places

```suggestion
        payload = jsonify({'notices': {}, 'msg': 'This feature is not implemented'})
```

File: frontend/tests/unit/api/notice/test_v0.py (hunk 2 above)
Comment (kristenarmes) [reply frontend/tests/unit/api/notice/test_v0.py:41]: saw this comment after I made my other comment related to this, is this just a note to self? I think either the basic msg or something more specific would work

Comment (B-T-D) [reply frontend/tests/unit/api/notice/test_v0.py:41]: Yeah they were notes to self when I was on a different branch making that little quality API PR, so couldn't just make the changes when I noticed them

File: frontend/amundsen_application/api/notice/v0.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    client = get_notice_client()
+    table_key = get_query_param(request.args, 'key')
+    response = client.get_table_notices_summary(table_key=table_key)
+    status_code = response.status_code
+    if status_code == HTTPStatus.OK:
+        try:
+            notices = json.loads(response.data).get('notices')
+            payload = jsonify({'notices': notices, 'msg': 'Success'})
+        except ValidationError as err:
+            raise Exception(f"Notices client didn't return a valid ResourceNotice object. {err}")
```
Comment (Golodhros) [inline frontend/amundsen_application/api/notice/v0.py:59]: This case won't send an error response?

File: frontend/amundsen_application/static/js/components/Alert/Alert.tsx [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+export const Alert: React.FC<AlertProps> = ({
+  message,
+  severity = NoticeSeverity.WARNING,
+  onAction,
+  actionText,
+  actionHref,
+  actionLink,
+  payload,
+}: AlertProps) => {
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/components/Alert/Alert.tsx:54]:
```suggestion
}) => {
```

File: frontend/amundsen_application/static/js/components/Alert/Alert.tsx [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+  const [showPayloadModal, setShowPayloadModal] = React.useState(false);
+  let action: null | React.ReactNode = null;
+
+  const handleSeeDetails = (e: React.MouseEvent<HTMLButtonElement>) => {
+    onAction?.(e);
+    setShowPayloadModal(true);
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/components/Alert/Alert.tsx:60]: this is missing the logging logic in https://github.com/amundsen-io/amundsen/blob/main/frontend/amundsen_application/static/js/components/Alert/index.tsx

File: frontend/amundsen_application/api/notice/v0.py (hunk 4 above)
Comment (B-T-D) [reply frontend/amundsen_application/api/notice/v0.py:60]: It seems the pattern from other endpoints (e.g. api/quality/v0.py, api/announcements/v0.py) is to log the error and raise the exception. I think this error should only be hit rarely--if it's just a bad request from the frontend, the notice client will return a more descriptive error response.

I updated the logging here to be more in line with what quality and announcements have.

**Comment by ozandogrultan** [comment]
Great work and addition! Do we any documentation on how to configure and use this feature? @B-T-D

**Comment by B-T-D** [comment]
I can try to add more documentation at some point. There's a bit of documentation in the form of docstrings in the code currently. 

Basically, you will need to have some separate service that is generating these notices. You then configure a concrete notice client that subclasses `BaseNoticeClient`. That notice client's job is to talk to whatever non-Amundsen service you're getting the notices from, and turn them into Flask responses that are understandable by the endpoint in `api/notice/v0.py`.

**Pull request merged**